### ISSUE TITLE
Gracefully draining Services/Endpoints

### ISSUE THREAD

**Issue by tssurya** [kubernetes/kubernetes#108523]
### What happened?

Currently when services or endpoints are deleted we remove the conntrack entries for UDP & SCTP services:
https://github.com/kubernetes/kubernetes/blob/88f97283393715c7fa870dd90be0fd0d0faad6bd/pkg/proxy/iptables/proxier.go#L736
https://github.com/kubernetes/kubernetes/blob/c5fbcd735d67a7cc209301dc690579fb0816e2ab/pkg/util/conntrack/conntrack.go#L132

This causes ABORT to never happen and hence no graceful termination. At the same time if we don't clean the conntrack entries and if users are re-using the src ports, then we'll have stale entries preventing the new entries from being created; example: https://github.com/kubernetes/kubernetes/issues/101607. SCTP is more like TCP and is stateful; so the expectation is for them to behave similarly; but today we are treating it like UDP?

So there are two ways forward:

1. When services are deleted we cleanup the entries and continue not allowing ABORTS or graceful terminations for already ESTABLISHED connections like we are doing now in kube-proxy.
2. We don't touch conntrack for connection oriented protocols and let the protocol take its natural course through the various states, maybe this will motivate people to use random source ports. This means even if services go away, ESTABLISHED connections continue to work until they get ABORTED and conntrack cleanup will happen naturally.

/cc @dcbw @danwinship @aojea @trozet

### What did you expect to happen?

The exact expectations or "right way of doing things" around this problem need to be discussed/figured out. Note that this item was brought up in the sig-network meeting on the 3rd of March 2022. We had a few ideas like:

Quoting @dcbw's meeting notes:
> deleting a Service object meant that the service was no longer guaranteed to work.
> Instead, to drain a service, the label gets changed and thus no endpoints, and new connections get rejected.
> But the Service sticks around and existing connections work until the backend pods are deleted, or the Service is finally deleted.

Since we didn't have a clear consensus, opening this issue where we can clearly document the agreed way forward. 

### How can we reproduce it (as minimally and precisely as possible)?

Same steps as here: https://github.com/kubernetes/kubernetes/issues/70020
The WORKAROUND2 mentioned in the above issue ^ was implemeted through https://github.com/kubernetes/kubernetes/pull/74073; but SCTP is stateful like TCP and since we nuke the entries, the SHUTDOWN/ABORT will never happen? We are not gracefully terminating.

### Kubernetes version

N/A

### Cloud provider

N/A

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
TCP behavior can not be changed, people expect that their applications terminate the connections gracefully, and stop doing that is a breaking change.
In UDP we don't know in which state is the connection, so we just simple drop it because the only signal we have is that the endpoint/service disappear, otherwise we blackhole traffic and services no longer work... I think that is the only solution we have :shrug: 
If SCTP handles sessions like TCP, it seems it should behave as TCP, but I'm a completely ignorant on how SCTP runs in production ... that is why I wanted to have a more realist e2e sctp test https://github.com/kubernetes/kubernetes/issues/103150 so we could decide with more data
@uablrek @ravens have much more visibility on SCTP environments , I can't find Per Andersson github handle, but it will be good to get feedback from people feedback in the field with this protocol

**Comment by uablrek** [comment]
Is graceful termination _really_ necessary when a service is deleted?

I mean, it's not the same as a pod termination which can be a part of a scale-in. When a used delete a service it should be because it should not be used anymore, e.g when unloading an app. A used that have a use-case where traffic _must_ be gracefully terminated when a service is deleted can gracefully terminate all endpoints _before_ deleting the service. However, I think those users are few. In most cases deletion of a service is part of an application unload or a brutal upgrade where disturbanced are expected anyway.

**Comment by uablrek** [comment]
Related https://github.com/kubernetes/kubernetes/issues/101968#issuecomment-843272392

Iptables and ipvs is different.

**Comment by aojea** [comment]
> Is graceful termination _really_ necessary when a service is deleted?

Heh, that as good question , I have to test that, if I connect to a service and I delete the service, do I keep the connection until the pod goes down?

Regarding endpoints, I was checking how other are doing and seems that haproxy has an option to explicitly indicate shutdown connection if backend is down http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#5.2-on-marked-down, is.something do we want to model in the API @thockin ?

**Comment by uablrek** [comment]
Please note that https://github.com/kubernetes/kubernetes/issues/101607#issuecomment-855461836 is closed;

> This issue is solved by a Linux kernel patch added in linux-5.14. Closing ...

**Comment by trozet** [comment]
> Is graceful termination _really_ necessary when a service is deleted?

Yes I believe it is. Otherwise you are bypassing the close states of a stateful protocol. This can lead to applications waiting to detect end states with timeouts, and retransmission failures. It is also avoiding things like TCP TIME_WAIT state which is there for a good reason. Imagine you have a service which is deleted and then immediately recreated, and the client and server re-use the same source/dest port (connection). One side of the connection may correctly try to initiate a new connection while the other end may still be transmitting packets from a previous connection. In other words, service tear down is not pod tear down, and the pods may retain previous connection state, even though that state has been flushed from the intermediary system.

Additionally, we don't know what other intermediary stateful systems the connection has traveled through. What if there is some external device in the connection path where the connection is not flushed. Will it have stale established connections sitting its table for the default SCTP/TCP timeout of 5 days because the connection was never closed properly?

IMO it's not up to kubernetes to decide how L4 protocols behave (other than UDP due to necessity) and should not make assumptions about connection states for stateful protocols.

Either way the current behavior is inconsistent. If the desire is for kubernetes to delete conntrack entries and ignore graceful termination, then conntrack needs to be flushed regardless of L4 protocol.

**Comment by uablrek** [comment]
@trozet That is all valid general arguments and applies to pod/endpoint termination. But services are different. Removing a service should only happen when you terminate/unload an app, in which case shutdowns doesn't matter. And;

> A user that have a use-case where traffic _must_ be gracefully terminated when a service is deleted can gracefully terminate all endpoint pods before deleting the service.

**Comment by khenidak** [comment]
Graceful termination is scaling pod replicas to zero with `SIGTERM` followed by configurable grace period followed by `SIGKILL` described in details here: https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-terminating-with-grace. That works with rolling forward, backward and eviction. Now even when a pod is deleted the buffers that was `send(2)` will still go out and connection mgmt according to keep-alive will still carry (at least on Linux)

Deleting a service is not a graceful action. And should not be used as such. If you are retiring an app (such as replacing it with another that may listen to different IPs/ports), then the steps are 1) scale to zero 2) wait for all connections to terminate 3) delete service.

/assign

**Comment by trozet** [comment]
@uablrek @khenidak what about removing a pod as an endpoint to a service. Is that a graceful termination case or not?

**Comment by danwinship** [comment]
Is there consensus that TCP is doing everything right already? What about the case of explicitly-specified source ports? There was some suggestion in the meeting that either (a) the reason SCTP doesn't work right is because when using SCTP people tend to explicitly bind a source port, and kube-proxy's model doesn't handle that, _OR_ (b) the reason our code treats SCTP and TCP differently is because we're trying to deal with explicitly bound source ports for SCTP but not for TCP. But it wasn't clear which of those two things people were claiming (or if the claim was correct). At any rate, it seems like we should handle source port reuse with both TCP and SCTP (if possible?)

There was discussion about a kernel fix in 5.14 but it's not clear exactly what it fixed or why we needed that. (The discussion there seems to be TCP-specific?) There has also been discussion of SCTP conntrack bugs. It is not clear if/when those bugs are fixed, or what the effect of them is, or whether we believe that any of our current code is an attempt to work around those bugs, successful or not.

The initial comment is focusing too much on SCTP messages and conntrack states, and not enough on end-user-visible behavior. If I was designing this from scratch and not constrained by what was actually possible, I'd say:

- If you have a TCP or SCTP service, and you delete the _Service_, then all existing connections should be forcibly closed (eg TCP RST).
  - If that's not possible / is possible but really hard, then we _could_ just blackhole them, since "deleting a Service while it has existing connections" is not really a normal operation.
- If you have a TCP or SCTP service, and you delete a Pod that is an endpoint of that service, then kube-proxy must immediately stop sending _new_ connections to that endpoint, and must not interfere with existing connections to that pod before the pod exits (ie, we should not delete conntrack entries for existing connections)
  - when the pod actually exits and is fully deleted (either because the termination grace period ends, or because you deleted it with `--force --grace-period 0`) _then_ any remaining connections to the service via that endpoint should be forcibly closed (and not just black-holed).

**Comment by uablrek** [comment]
> What about the case of explicitly-specified source ports?

### TCP

I would say it "should work" with https://github.com/torvalds/linux/commit/e15d4cdf27cb0c1e977270270b2cea12e0955edd that was added in linux-5.14. I have tested it manually as described in https://github.com/kubernetes/kubernetes/issues/101607#issuecomment-1059911699. Even when the client constantly re-connects using the same port, the stale conntrack entry expires after 120s, the entry is removed and the next connect is succesful.

**Comment by trozet** [comment]
If a TCP service is deleted today obviously the ports are not cleared and the connection continues to work. If we make a change where TCP conntrack entries are flushed, then a client trying to send the next TCP packet on the now deleted conntrack entry would force the kernel to send a TCP RST. Which sounds like the desired behavior. We need to change TCP to also be flushed on service deletion, ass well as document this behavior.

I'm still wondering about endpoint deletion:
https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1061225334

**Comment by uablrek** [comment]
Note, TCP RST is a response. If a client is silent (not sending anything) it will not receive any RST's. This is different from a FIN.

I remind about the current behavior when a service is deleted (https://github.com/kubernetes/kubernetes/issues/101968#issuecomment-843272392);

**proxy-mode=iptables** - The connection lives on as if nothing has happened. It is up to the application to detect the problem.

**proxy-mode=ipvs** - The connection is black-holed. Communication just stops. Application can not count on a connection close or reset but must detect the problem with keep-alive or timeouts.

Neither of these is the desired behavior. For iptables I think that clearing conntrack will trig a RST response. For ipvs it is not enough. If the VIP is removed from the `kube-ipvs0` interface (which it will not if many services use the same LBIP) you _might_ get a no-route-to-network RST, but I am unsure.

**Comment by uablrek** [comment]
> I'm still wondering about endpoint deletion:

If the service has no selector the user has taken the responsibility of handling this. It can mimic the (rather complex) graceful termination of K8s, or do something else. User handled endpoints can vary a lot, like beeing outside the cluster for instance, so IMHO K8s should not interfere or "help" in any way.

If an endpoint is removed because the pod has become not-ready its different. I don't think one can assume degrees of not-readiness, so it should be assumed that the pod is dead, and IMO it should be handled as a deleted service.

**Comment by trozet** [comment]
I'm still not quite clear on this. Consider the following example:
client pod ---> service ---> server pod

The service is using a label selector to determine server pod is an endpoint. Client has a currently established TCP/SCTP connection to server. Now, the server pod is updated and the label is removed that selects it as an endpoint to this service. Should the connection continue to function until graceful termination? Or should that conntrack entry be flushed?

**Comment by palonsoro** [comment]
From some experience "on the field", it is kind of usually expected that the readiness probes mean "ready to accept new connections", not "ready to accept new and keep existing connections".

Let's consider this example use case: some application whose thread pool is configured such that it can handle up to 400 TCP connections. A readiness probe may return "true" or "false" based on whether the application is already handling those 400 connections or not (among other criteria) so that, if one replica pod is handling the 400 TCP connections but the other replica pod isn't, the "too busy replica" would be removed from endpoint list but not get any new connection, but still be able to process the current ones and eventually free some slots, report ready again and return to endpoint list.

I don't see this equivalent to a more "intentional" action like deleting a service on purpose, which to me sounds like unplugging a wire.

**Comment by danwinship** [comment]
> Now, the server pod is updated and the label is removed that selects it as an endpoint to this service.

I'm not sure that's an action we expect users to take, so I'm not sure there's any specific expected behavior...

Kube-proxy doesn't make any attempt to track you doing this; if a previously-working endpoint stops matching the service's selector, then kube-proxy immediately forgets about the endpoint (ie, as though you had deleted it with `--force --grace-period 0`).

**Comment by uablrek** [comment]
I don't know a way to forcefully break a TCP connection when the peers are silent. It would be like a man-in-the-middle attack. A ICMP `Destination Unreachable` can be sent by any network element along the path, but is also a response.

**Comment by uablrek** [comment]
> You can't send an "unsolicited" RST?

Actually I become unsure. I have only seen it as an OS response. The OS should be able to send a unsolicited RST, but I don't know if it ever does. I tested to setup a tcp connection between peers on 2 vms and did a `kill -9` on the server, but the OS close the connection with FIN.

**Comment by ravens** [comment]
Regarding SCTP, from a user perspective (working for a telco software company, heavily users of SCTP) it would be great to gave a graceful disconnection when endpoint are disappearing; when services are removed, well I don't think there is any exceptions regarding the behavior, imo.

To give a more concrete example of what does it means for a telco use case for example, consider this example on one of our eNodeBs (4G base stations, go-based SCTP application behind a service on a k8s cluster, node are ubuntu 20.04 with 5.13.0-30-generic kernel, most defaults everywhere):
- start my pcap capture at T=0
- I am killing my SCTP pod
- endpoint is changing, pod is restarted etc.
- service IP is the same
- eNB does not get notification of anything, but SCTP keep alive packets are now flowing again to the pod (packet highlighted in blue), i.e. after 30s. 
- a phone tries then to connect to the antenna, leading the eNB to send a couple of message (S1AP, which is running over the SCTP socket)
- the SCTP traffic is ignored on the target pod bc no actual socket opened SCTP from pod perspective 
- after a certain times, the eNB consider the lack of message response, and disconnect/reconnect
- telco service is now functioning back to normal

![image](https://user-images.githubusercontent.com/104708/158072503-6136b582-e8c8-4ef2-b3df-0bc1ac5f8149.png)

**Comment by uablrek** [comment]
> I am killing my SCTP pod

Gracefully?

Linux kernel sctp (lksctp) or user-space?

What are you expecting of K8s when this happens (if you could get anything)?

When an sctp socket is closed by the kernel because the process is killed I would expect the kernel to terminate (abort) the sctp association, similar to closing a TCP-connection when that happens. If the kernel doesn't do that, there is not much K8s can do.

**Comment by ravens** [comment]
@uablrek lksctp using C bindings from go. ideally I would have expected some ABORT coming back to the eNB. I will try do redo a more controlled experiment, i.e. within k8s or not and compare the behavior. But I think the idea was to illustrate the kind of issues in the context of the discussion, i.e. the original issue: 
`This causes ABORT to never happen and hence no graceful termination.`

**Comment by DenisScherbakov** [comment]
Hello guys, 
I'm joining to this discusstion quiet late. I was involved into issue #70020 which was resolved by PR #74073.
We work in telecom area and widely use SCTP as well. 
Iit is not clear for me what kind of a problem we introduced by flushing conntrack entries? Lack of possibility to send ABORT to peer is a problem, right?

Let me explain why we thought flushing of conntrack entries is a good idea:

1.  graceful termination via ABORT just moves conntrack entry to [CLOSED] state which has time-to-live 10 secs. Let's imagine a case where SCTP in k8s takes a role of server and we have a peer client which communicates with a server via LoadBalancer service.  In our experience if server pod is gracefully closed and it sent ABORT then typical reaction on client's side is to immediately try to recover connection with using same source port (as it was discussed in this topic it is a usual case for SCTP users to bind to specific port). As result new INIT comes to k8s whlie previous [CLOSED] conntrack entry still there and INIT handled in a wrong way -> k8s tries to route INIT to destination which does not exists anymore. Peer is not able to recover sctp association, and the only way is to "teach" client to re-establish association not immediatelly but after 10 sec. I think this is a wrong approach. Why client should bother about server's realization and conntrack specific timing behaviour? If client just follows SCTP RFC protocol standard then it has a freedom to recover connection immediatelly. 

@ravens as I can see in your wireshark example there is a 10 sec gap between ABORT and next INIT? was it done intentionally just because you're aware of such conntrack behaviour? What if client re-creates an association earlier?

2. a key feature of k8s is to provide high availability for applications in error cases. If container is crashed or deleted with --force --grace-period 0 options then application might not be able to close associations gracefully. K8S should recover application but without conntrack flushing system remains unhealthy. 

So if I would chose between "lack of graceful termination" and "unhealthy application" then I vote for "lack of graceful termination".

**Comment by ravens** [comment]
@DenisScherbakov the 10s was not intentional - this is an hardware eNB that has an embedded stack acting as a SCTP client I don't really control. I need to to a more controlled experiment and take a pcap at the edge of the cluster.

**Comment by aojea** [comment]
> 1. In our experience if server pod is gracefully closed and it sent ABORT then typical reaction on client's side is to immediately try to recover connection with using same source port (as it was discussed in this topic it is a usual case for SCTP users to bind to specific port).

> Clients bind the source port more often than TCP clients

I see the problem, so SCTP in the wild doesn't behave as TCP , that means the hypothesis based on the academic definition of stateful and stateless protocols will break the actual SCTP users ... 
but still seems to be scenarios that people will benefit of being able to choose between shutdown connections on close, there were several request to implement that on TCP too for typical active/passive deployments ... 

If we consider adding an option to the Service to define this behavior, like a bool `GracefulShutdownSession`  there are several things to have into consideration:
- the option is per protocol, so it can only go in the Ports field of services
- each protocol has different defaults: UDP: true, TCP false: SCTP: ??
- does UDP will ever use GracefulShutdownSession false?
- ...

```go
type ServicePort struct {
	// GracefulShutdownSession allows removing endpoints from a service without forcefully close the sessions,
	// Defaults are: TCP: true, UDP: false, SCTP: true (xref #108523) 
	GracefulShutdownSession *bool
}
```

so an user can define something like:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: svc1
  namespace: default
spec:
  ports:
  - name: https
    port: 443
    protocol: TCP
    targetPort: 443
    gracefulShutdownSession: false
  - name: sctp
    port: 8001
    protocol: SCTP
    targetPort: 80 
    gracefulShutdownSession: true
  - name: udp
    port: 5001
    protocol: SCTP
    targetPort: 5001
    gracefulShutdownSession: true
```

WDYT?

**Comment by teiclap** [comment]
In SCTP the graceful shutdown is not ABORT.
From SCTP perspective it's much better not to try to do any graceful operation.

**Comment by aojea** [comment]
> In SCTP the graceful shutdown is not ABORT.
> From SCTP perspective it's much better not to try to do any graceful operation.

@teiclap  can you please elaborate? 
I can see in the RFC that the protocols support both shutdown modes https://datatracker.ietf.org/doc/html/rfc4960#section-3.3.7 

We should explain the reasons, and this also will benefit to educate people without experience in this protocol (like me :) )

**Comment by teiclap** [comment]
A graceful shutdown is meant not to loose any data, this is accomplished in SCTP by means of SHUTDOWN procedure. ABORT doesn't guarantee that data is flushed. 
Let's try and see what happens when you remove a SCTP service is some different modes:
1. SCTP server crashes. The remote peer will notice it by exploting the retransmission counter, when expired the peer will drop the association itself. Since the server has crashed there's not way for sending any ABORT or whatever. The remote peer will try sending data for a while.
2. SCTP server works but the service is closed and conntrack is cleaned. From peer perspective is the same as 1.

We see that the problem is how the remote peer experiences the problems and how they have tuned the internal timers for detecting the peer's outage. This is a known problem with SCTP and the way rfc4960 recommends timers doesn't really work well in such cases as SCTP assumes that a problem in the network is more easy to happen than a crash in the peer, so that the remote peer may be waiting for unreasoneable long time before understanding that the association is gone.

Case 2 can be solved if conntrack cleaning is coordinated in some way, meaning that the application using SCTP does close the service in a proper way, shutting down all Association before conntrack is cleaned.

**Comment by thockin** [comment]
Capturing some notes from sig-net discussion.

We have a lot of things twisted up here.  We've never really written out the "state machine" here specifically, what SHOULD happen to {new | existing} flows when:
1) endpoint becomes unready
2) endpoint becomes terminating
3) endpoint is removed 
4) endpoints(lice) becomes terminating (e.g. has a finalizer)
5) service becomes terminating (e.g. has a finalizer)
6) service is deleted

Realisitically, we might get different answers per-protocol here, and we should be cognizant that not every implementation is as flexible as every other, so we should spec carefully.

Here's an off-the-cuff take, remember this is what I think it *should* do, not what it DOES do or even CAN do.

1) endpoint becomes unready
  - New flows should not be established
  - Existing flows should be untouched
2) endpoint becomes terminating
  - New flows should not be established (unless no other non-terminating endpoints exist)
  - Existing flows should be untouched
3) endpoint is removed 
  - New flows should not be established
  - Existing flows should:
    - TCP: RST/ICMP/something immediate
    - UDP: break conntrack
    - SCTP: whatever SCTP experts say is best
4) endpoints(lice) becomes terminating (e.g. has a finalizer)
  - handle it as if the slice was deleted
5) service becomes terminating (e.g. has a finalizer)
  - New flows should not be established
  - Existing flows should be untouched
6) service is deleted
  - New flows should not be established
  - Existing flows should be broken

Does anyone have strong disagreement?  I know (3) is the main point of contention, and I feel like it may be OK to call that "implementation defined".  (4) is an omission in ALL of our controllers.  (6) is not something I think I have ever thought about, but it feels like a service being deleted is ~ the same as a server crashing.  Whether that should get a RST or not is debatable.  I could agree to "implementation defined".

I think I am mostly restating others's remarks, but here's where I landed.

> when the pod actually exits and is fully deleted ... then any remaining connections to the service via that endpoint should be forcibly closed (and not just black-holed).

The problem here is that the removal of an endpoint IS NOT THE SAME as the pod being killed.

@trozet said

> The service is using a label selector to determine server pod is an endpoint. Client has a currently established TCP/SCTP connection to server. Now, the server pod is updated and the label is removed that selects it as an endpoint to this service. Should the connection continue to function until graceful termination? Or should that conntrack entry be flushed?

Exactly.  If we flush conntrack, TCP will get a RST on the next client->server packet (I think?), UDP will pick a new backend on the next packet, SCTP will ... what?

If we do not flush, then TCP will stay alive until the endpoint terminates and then send RST, UDP stay alive until the endpoint terminates, and then will blackhole all subsequent connections, and SCTP will stay alive until the endpoint terminates.

UDP in particular is a problem because once the endpoint is gone, we no longer are paying any attention to that conntrack record.  If the pod actually does go away, we don't know about it.  So we HAVE to flush UDP in case (3).

Unless I am mistaken, flushing or not flushing TCP is a semantic choice - do we want to leave those connections alive when the endpoint is not "in" a Service.  

As for SCTP... does it get something like a RST when the connection is flushed?  Given all the talk of user-space SCTP, I'm guessing not (not guranteed anyway).

**Comment by uablrek** [comment]
> As for SCTP... does it get something like a RST when the connection is flushed? Given all the talk of user-space SCTP, I'm guessing not (not guranteed anyway).

Yes, ABORT. That was a (the?) reason for _not_ loading the sctp kernel module. The user-space sctp stack used AF_PACKET (raw) sockets (I have tried to point out that there are [better options](https://github.com/sctplab/usrsctp/issues/631) nowdays) but a raw socket _clones_ packets, it does not redirect them, so if the sctp module (lksctp) is loaded the _kernel_ will send ABORT on sctp associations handled by the user-space stack.

When conntrack entries are flushed the ABORT will not be sent automativally by the kernel and lksctp users [might want that](https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1066151442). But from what I understand the problem with re-esablish a new association if the conntrack entries are not flushed is more devastating.

**Comment by bowei** [comment]
@thockin : aren't individual endpoint states defined by their conditions, not finalizers?

Use of finalizers as a state signal for Service is an interesting idea, although it couples
resource lifecycle with connection behavior in a subtle way.

```text
   [no_endpoint]

     |  Pod created
     V

   [endpoint_not_ready]
     !ready, !serving, !terminating

     |  Readiness healthcheck OK
     V

   [endpoint_ready]
     ready, serving, !terminating

     |  Pod has deletion timestamp
     V

   [endpoint_terminating]
     ! ready, serving, terminating

     |  Pod is deleted
     V

   [no_endpoint]
```

**Comment by teiclap** [comment]
> When conntrack entries are flushed the ABORT will not be sent automativally by the kernel and lksctp users https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1066151442. But from what I understand the problem with re-esablish a new association if the conntrack entries are not flushed is more devastating.

Sending ABORT when the SCTP host dies is not part of the protocol specification.
It's up to the protocol implementation to instantiate a SIGKILL handler that does ABORT all the existing Associations before dying and it's not guaranteed that such ABORTs will ever reach the peer because of flushing conntrack.

In my opinion shutting down a service should be done in a coordinated way, that is informing the application about and giving it the time for releasing all resources, that involves all the SCTP Associations. Any other way has potentially the same effect as unplugging the network cable.

**Comment by aojea** [comment]
> 1. endpoint becomes unready
> New flows should not be established
> Existing flows should be untouched

+1

> 2. endpoint becomes terminating
> New flows should not be established
> Existing flows should be untouched

+1 though this is still beta, [are we missing something for GA?](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1672-tracking-terminating-endpoints) @andrewsykim 

https://github.com/kubernetes/kubernetes/blob/a504daa0484d35d07f144e3f65f5a481570e96ec/pkg/features/kube_features.go#L491-L494 

> 3. endpoint is removed
> New flows should not be established
> Existing flows should:
> TCP: RST/ICMP/something immediate
> UDP: break conntrack
> SCTP: whatever SCTP experts say is best
 
+1  with "implementation defined" behavior, or we just say "Close established sessions to the removed endpoint" and let everybody define what they mean by close :smile: 

> 4. endpoints(lice) becomes terminating (e.g. has a finalizer)
>  is an omission in ALL of our controllers

yeah, that is not a thing https://github.com/kubernetes/kubernetes/issues/108267#issuecomment-1059224270

> 5 service becomes terminating (e.g. has a finalizer)

endpoints/slices remains alive, not even terminating, should the endpoint slices controller react and sets the slices as terminating? this is related to the issue opened by @robscott  https://github.com/kubernetes/kubernetes/issues/108779

If we modify the slices behavior to follow the service status,  5 will become 2

> 6 service is deleted

 the endpoints/slices are automatically removed, so 6. will be 3. and there is nothing else to do , it also matches [Lars' comment](https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1059418199)

I think that endpoints/slices are the ones that should "lead" the behavior, so we can control the state from the slices controllers with the conditions status.

**Comment by thockin** [comment]
@bowei 

> @thockin : aren't individual endpoint states defined by their conditions, not finalizers?

Individual endpoints (as represented in Endpoints and EndpointSlice) don't get finalizers.  We have to rely on the metadata express in-band in those objects.

So I am not sure what youi are asking?

@teiclap 

> In my opinion shutting down a service should be done in a coordinated way, that is informing the application about and giving it the time for releasing all resources, that involves all the SCTP Associations. Any other way has potentially the same effect as unplugging the network cable.

SHOULD is fine, but we need to define what happens when it _isn't_ clean.  Saying "the same effect as unplugging the network cable" is fine in the abstract, but in practice a) that's not what we do today (have to be careful of breaking changes) and b) how much wiggle room should we leave for implementations.

@aojea 

> 5 service becomes terminating (e.g. has a finalizer)
>
> endpoints/slices remains alive, not even terminating, should the endpoint slices controller react and sets the slices as terminating? this is related to the issue opened by @robscott #108779
> 
> If we modify the slices behavior to be Terminating if the service has a finalizer, then 5 will become 2

We can do that for slices we control, but we still have to define the behavior for slices we don't control, so I think it needs to be called out.

**Comment by thockin** [comment]
> If we flush conntrack, TCP will get a RST on the next client->server packet (I think?)

I can't empircally prove this.  When I manually flush conntrack while a TCP session is open to a service, the session continues and the conntrack record is restored.  This is not what I expected.  Anyone have more depth on this?

**Comment by aojea** [comment]
> I can't empircally prove this. When I manually flush conntrack while a TCP session is open to a service, the session continues and the conntrack record is restored. This is not what I expected. Anyone have more depth on this?

the iptables DNAT rule is still there, right?

**Comment by andrewsykim** [comment]
> +1 though this is still beta, [are we missing something for GA?](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1672-tracking-terminating-endpoints) @andrewsykim

Just one clarifying point about EndpointSliceTerminatingCondition w.r.t this issue. The API now supports terminating endpoints but we don't do anything in kube-proxy to handle the terminating condition yet. ProxyTerminatingEndpoints allows fallback to those endpoints only when there are terminating endpoints which is addressing a slighly different issue. An alternative approach to ProxyTerminatingEndpoints like https://github.com/kubernetes/kubernetes/pull/108574 could potentially help with this issue but it expands the Service API in an ugly way.

**Comment by aojea** [comment]
> > I can't empircally prove this. When I manually flush conntrack while a TCP session is open to a service, the session continues and the conntrack record is restored. This is not what I expected. Anyone have more depth on this?
> 
> the iptables DNAT rule is still there, right?

hmm the conntrack entry I see is ESTABLISHED but UNREPLIED

```
tcp      6 202 ESTABLISHED src=172.18.0.3 dst=10.96.141.142 sport=54988 dport=80 [UNREPLIED] src=10.96.141.142 dst=172.18.0.3 sport=80 dport=54988 mark=0 use=1
```

1. create a persistent connection

```
nc -v 10.96.141.142 80                                                                                                                                                                                    
Connection to 10.96.141.142 80 port [tcp/http] succeeded!
...
```

2. delete service and conntrack entry
```
kubectl delete service/test
conntrack -L -p tcp -d 10.96.141.142
tcp      6 235 ESTABLISHED src=172.18.0.3 dst=10.96.141.142 sport=54988 dport=80 [UNREPLIED] src=10.96.141.142 dst=172.18.0.3 sport=80 dport=54988 mark=0 use=1
conntrack -D -p tcp -d 10.96.141.142
```

3. it seems it will eventually timeout
```
tcp   CLOSE-WAIT 0      57                                                                             172.18.0.3:54988             10.96.141.142:80        users:(("nc",pid=6994,fd=3)) timer:(on,1min19sec,9)                    
```

**Comment by thockin** [comment]
initial:
```
$ sudo conntrack -L 2>/dev/null | grep 10.64.2.20 | grep 224.27
$
```

telnet to svc:
```
$ sudo conntrack -L 2>/dev/null | grep 10.64.2.20 | grep 224.27
tcp      6 431997 ESTABLISHED src=10.64.2.20 dst=10.0.224.27 sport=53144 dport=80 src=10.64.3.28 dst=10.64.2.20 sport=9376 dport=53144 [ASSURED] mark=0 use=1
$
```

clear conntrack:
```
$ sudo conntrack -F
conntrack v1.4.4 (conntrack-tools): connection tracking table has been emptied.
$
```

verify:
```
$ sudo conntrack -L 2>/dev/null | grep 10.64.2.20 | grep 224.27
$
```

hit enter in telnet (get HTTP response, closes connection):
```
$ sudo conntrack -L 2>/dev/null | grep 10.64.2.20 | grep 224.27
tcp      6 117 TIME_WAIT src=10.64.2.20 dst=10.0.224.27 sport=53144 dport=80 src=10.64.3.28 dst=10.64.2.20 sport=9376 dport=53144 [ASSURED] mark=0 use=1
$
```

**Comment by trozet** [comment]
I was doing my test with OVN-Kube on a single node. I think from the discussion on this issue OVN-Kube CNI is going to go with:
1. Service deletion - remove NAT entries and flush conntrack for any protocol
2. endpoint deletion - do nothing for TCP/SCTP.  UDP flush conntrack.

If this issue gets resolved later with a different behavior we will look at moving to that, but in the interim we need a solution with behavior that is consistent across connection oriented protocols.

**Comment by aojea** [comment]
>  (as it was discussed in this topic it is a usual case for SCTP users to bind to specific port).

@DenisScherbakov is this a fact or just a personal perception?

> From SCTP perspective it's much better not to try to do any graceful operation.

@teiclap this is confusing me, because in another comment you describe the graceful operation for SCTP

> Case 2 can be solved if conntrack cleaning is coordinated in some way, meaning that the application using SCTP does close the service in a proper way, shutting down all Association before conntrack is cleaned.

is that another thing like the binding source port, that SCTP applications don't use to  implement this behavior, hence we can be sure the connection can be closed abruptly?

**Comment by thockin** [comment]
@bowei Yeah, I see that behavior when I test.  "helpful".  Sigh.

**Comment by teiclap** [comment]
> 

@aojea You cannot do any graceful shutdown without involving the whole stack on top of SCTP.
I guess that the requirement is to give to the remote peer a chance to understand that the service has been closed.
You are assuming here that an ABORT message will reach the remote peer, this is not guaranteed as ABORT will not
be replied but let's continue assuming that it will arrive once sent.
Even sending such ABORT when dying is not part of SCTP, it's part of the specific implementation thus it's not a requirement for K8s.
What is being observed earlier, that is ABORT was generated spontaneously, it's a side effect of having a different SCTP protocol stack reacting when hit by unknown traffic. In such case Linux Kernel SCTP is fed with the traffic that was previously routed to the service.
A generic solution that doesn't depend on the type and configuration of the host OS (i.e. it will work on Linux, BSD, Windows) will require an implementation of the SCTP protocol stack, possibly in the user space running in a POD, that will react to being killed (SIGTERM, SIGPIPE) by sending ABORT to all the remote peers and, at the same time, a NAT agent that will close the tracking service after that the service is closed. 
I don't like the implementation of a conntracker that parses all the Control Chunks (thus all SCTP packets) for guessing what to do, I'd rather adopt a different approach but this is another story.

As a conclusion, I think that K8s does things properly as it is, that conntrack closes the tracking properly and that the missing point is in the implementation of SCTP stack (if any)

**Comment by danwinship** [comment]
> > As for SCTP... does it get something like a RST when the connection is flushed? Given all the talk of user-space SCTP, I'm guessing not (not guranteed anyway).
> 
> Yes, ABORT. That was a (the?) reason for _not_ loading the sctp kernel module.

You're mixing up two entirely unrelated things.

The reason for not loading the SCTP kernel module is that doing so will kill _active_ SCTP connections (because the kernel sees packets it doesn't know about, addressed to an IP that it owns, so it will send responses back to the client telling it "I don't know anything about this connection you seem to be sending packets to", which I guess are ABORTs; this is no different from what would happen if someone were implementing TCP in userspace and the kernel TCP stack saw the packets.)

What Tim was suggesting (I think) is that if someone is using userspace SCTP, then there is less possibility for graceful termination, because once the pod is killed or stops receiving packets, then there's no else left who knows anything about that SCTP connection and can try to clean it up. But that's not _entirely_ true, because even when you're using userspace SCTP, the kernel is still handling SCTP conntrack, so kube-proxy's conntrack-cleanup-up (or not-cleaning-up) actions are still relevant in at least some cases.

**Comment by qabpea** [comment]
I am coming in late into this exchange.

TCP, UDP, and SCTP all use 5-tuples to express a connection/flow
TCP and SCTP has state machines defined for setting up and tearing down connections.

There is a big difference in what the expected behavior based on if an ephemeral port is used as the source port or if a "static/configured" port number is used.
Take TCP as an example, TIME_WAIT is there to protect the IP stack from old packets in transit and to ensure that a newly established connection will not be old junk.
This is not a problem at all if the initiating party change the source port when setting up a new connection.
The connection tracker can keep the entry for some time before it is dropped, the chance for a duplicate 5-tuple is very low (better safe than sorry)
If the source port is locked, well then the existing 5-tuple entry in the conntracker is stopping the client from reconnecting with the server. 
There is no way that the proxy can know which semantics that is expected for either TCP or SCTP, we can make an educated guess of what the default behavior should be, but it will never work for bot cases unless we add a hint to the service which behavior is expected for this service/port.
@teiclap 100% agree, to parse all packets is expensive, but that is needed for statefull management.
Is the solution to be able to select stateless | statefull connection management for TCP and SCTP and make it possible do define in the service what the expected behavior is? This will of course then make everything timer driven 
which is what is already used in some TCP LBs with very short lived connections and high setup rate (6s timer for teardown)

**Comment by thockin** [comment]
> What Tim was suggesting (I think) is that if someone is using userspace SCTP, then there is less possibility for graceful termination, because once the pod is killed or stops receiving packets, then there's no else left who knows anything about that SCTP connection and can try to clean it up. But that's not entirely true, because even when you're using userspace SCTP, the kernel is still handling SCTP conntrack, so kube-proxy's conntrack-cleanup-up (or not-cleaning-up) actions are still relevant in at least some cases.

I think what I was saying was that deleting the conntrack record for a TCP session with DNAT would result in a RST, but I am wrong.  Since that would happen (usually) on the client-node, the next client packet would actually go on the wire (since iptables and IPVS are connection-time hooks (well iptables is, I think IPVS is too)) and presumably go into a black hole.  So my whole premise was flawed.  Without that conntrack record there is nobody to send that RST back.  Ditto ABORT.  Now, it turns out that Linux "helpfully" recovers the conntrack record ANYWAY (for TCP) so... moot and double moot.

**Comment by thockin** [comment]
What we really want is a one-time "poison this TCP session" trick.  Once a packet hits it, the poison HAS TO go away (hence an iptables rule is no good).

**Comment by bowei** [comment]
@thockin you could synthesize a bad packet in the flow -- it might make the conntrack entry get thrown out...

**Comment by teiclap** [comment]
@qabpea Here many people assumes that SCTP is a sort of variant of TCP, but it's quite a different beast.
The approach of conntrack in Linux and the similar mechanism in BSD relies in having a copy of the internal SCTP Host state machine for being capable of properly handling the connections.
At first, such approach doesn't take into account the multihoming feature of SCTP, meaning that a set of connections are belonging to the same Association and conntrack will see the initiation and termination only on one of them. An attempt in Linux conntrack has been made considering HB chunks but it doesn't help solving the case when the association is removed.
At second, one of the characteristics of SCTP is that it runs path verification continously, meaning that it's granted that there will be at least one HB chunk being sent on each path being part of an association every time HB timer expires.
At last, in order to understand whether an SCTP packet contains an INIT chunk, it's just enough to look at the SCTP header as the SCTP packet containing chunk is unique because it has Vtag=0.
The ideal conntrack for SCTP can be based purely on timer and close the connection when the timer expires.

For being polite and advertise the remote peer that the Association is gone, it's the SCTP Host implementation that SHOULD be designed in a way that when the service is closed (via SIGTERM in Unix machines), the SCTP does try to send ABORT towards all the existing Association. Being purely based on timer, conntrack will close the connection tracking at timeout.

From a clean implementation perspective, it's by far a better choice to run SCTP in user space and having no SCTP running in the kernel, and by the way no services in the kernel need SCTP.

**Comment by qabpea** [comment]
@teicap yes I am painfully aware of that TCP and SCTP are very different protocols, I implemented support for both SCTP and SIP in load balancers when I worked at Ericsson :-).
Multihoming in SCTP and MP-TCP is tricky when you do load balancing, you can not be sure that the packets for each direction/flowlet pass through the same device any longer and so on, combine that with direct server return and things get complicated.

For the issue at hand here, I believe it makes sense to look at where the session is instantiated from. If this is done by a POD in the cluster, then all packets in both directions will typically pass through the kernel of that node and the conntracker will have a chance to manage the sessions.

I agree with you regarding SCTP and user space, but that makes it almost impossible to provide SCTP support in K8s services unless the Proxy DP is also out of kernel.

To summarize to see if I understood you: It is better to have a timer based connection tracking of sctp associations with a fairly short timer. It would be good if the conntracker can manage the multihomed aspects of sctp. The protocol it self will guarantee to send packets across "each path".

This is what always concerned me, I can understand that sctp will send chunks to each destination, but how to cover different paths between the endpoints. Also, if no path exist towards a destination for a duration of time and the conntracker cleans up the entries, how will it then work when the path is unblocked again?

**Comment by teiclap** [comment]
> To summarize to see if I understood you: It is better to have a timer based connection tracking of sctp associations with a fairly short timer. It would be good if the conntracker can manage the multihomed aspects of sctp. The protocol it self will guarantee to send packets across "each path".
> 
> This is what always concerned me, I can understand that sctp will send chunks to each destination, but how to cover different paths between the endpoints. Also, if no path exist towards a destination for a duration of time and the conntracker cleans up the entries, how will it then work when the path is unblocked again?

@qabpea The conntracker is local, thus is SCTP fails sending HB in the expected time, the SCTP host is gone. 
If it's the peer that doesn't send HB, it doesn't matter as it will be always the local SCTP refreshing the tracker by means of the same 4-uple.
You can cover paths towards different destination by trusting outgoing traffic.
There's work ongoing at IETF, please have a look at 
https://www.ietf.org/archive/id/draft-porfiri-tsvwg-sctp-natsupp-02.txt

**Comment by qabpea** [comment]
@teiclap thanks for the rfc reference, I will read it over the weekend

**Comment by thockin** [comment]
@qabpea:

> @thockin when you mentioned "poisoned tcp session" is that in combination with old conntrack record and have the tracker send a RST instead of re-establishing the flow or something like that?

> The "proxy" functionality could potentially know if the endpoint/POD has terminated and only remove all connection entries in such a case, but this seems like adding quite a bit of new semantics to the services, proxy and endpoints

It's a hypothetical - I don't know how it would actually be implemented.  Why is this important to me?

Here's the common case:

* server pod is running
* client is connected via the service IP
* conntrack record exists
* server pod becomes terminating (graceful)
* endpoint is either removed from Endpoint{s,Slice} or marked `terminating`
* conntrack record still exists
* client connection still works
* no new connections are routed to the server
* server pod actually exits
* endpoint is removed from Endpoint{s,Slice}
* closing the process should have terminated the TCP session, and the client would be cleanly closed

In that case, none of this matters.  But there's a different case:

* server pod is running
* client is connected via the service IP
* conntrack record exists
* someone changes the labels on the server pod or service, such that the server pod is no longer selected (e.g. a quarantine)
* endpoint is removed from Endpoint{s,Slice}
* no new connections are routed to the server
* server pod is still running
* conntrack record still exists
* client connection still works

Is that OK?  What did the user INTEND by removing the endpoint?

IMO the "least surprise" behavior would be that removing an endpoint from Endpoint{s,Slice} would break the connection.  For UDP we achieve this by clearing the conntrack record.  I don't know how to implement it for TCP.

Then the follow-on case:

* server pod is running
* client is connected via the service IP
* conntrack record exists
* someone deletes the service entirely
* all endpoints are removed from Endpoint{s,Slice}
* no new connections are routed to the service at all
* server pod is still running
* conntrack record still exists
* client connection still works

This seems wrong, too, but also seems less likely to be real.

So TL;DR: What I think *should* happen for TCP maybe can't (at least not in iptables/ipvs) and leaving that as implementation-defined is acceptable.

For SCTP, I kind of lost the thread - is the result that we should treat it like UDP and clear conntrack, or leave the conntrack on the off-chance that the server will send an ABORT voluntarily (seems unlikely to me)?

**Comment by uablrek** [comment]
> * someone changes the labels on the server pod or service, such that the server pod is no longer selected (e.g. a quarantine)

> Is that OK? What did the user INTEND by removing the endpoint?

Please see https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1062747798.

A server temorary makes itself "not-ready" to avoid new connection, while still serving the existing ones. This is the best use-case I have seen. What the user expects in other cases (changing labels, removing endpoints) I think varies.

**Comment by qabpea** [comment]
@thockin 

I tend to look at the service and connections generated towards the pods as part of a contract.
I agree with you regarding case #1 the connections should stay in place and gracefully shutdown.

Regarding both case #2 and #3 I see it as the contract has been broken and that all connections should be severed directly as a  consequence of this.

For the SCTP, I will read through the RFC proposal tomorrow. There are a lot in SCTP like the multihomed aspects that is not a good fit to k8s to start with.
//Per

**Comment by thockin** [comment]
> A server temorary makes itself "not-ready" to avoid new connection, while still serving the existing ones. 

This is different than the cited condition, which is specifically "this endpoint is no longer part of the endpoint set".  We can now distinguish the difference between "not ready" and "not part of the set".

**Comment by qabpea** [comment]
@teiclap the draft RFC is good and puts a spot light on many of the problems we see here.
I assume that we can disregard multihoming in this discussion and only look at single path.

My first reaction is that the proposed behavior for a SCTP NAT is so different and that SCTP actually need a completely different mechanism then what the proxy and services offer. That said, this will not happen and we have what we have.

**Comment by uablrek** [comment]
> My first reaction is that the proposed behavior for a SCTP NAT is so different compared to TCP and UDP that SCTP actually need a completely different mechanism then what the proxy and services offer. That said, this will not happen and we have what we have.

I _totally agree_ on this. I think more advanced SCTP users (telco) _must_ write customized proxiers. Probably with [kpng](https://github.com/kubernetes-sigs/kpng). I made a https://github.com/kubernetes/kubernetes/issues/92369#issuecomment-650034563 about using a customized SCTP proxier at the birth of kpng. If this is kind of accepted then the `kube-proxy` can target the sctp-as-packet-based-tcp users (i.e. _not_ telco).

**Comment by teiclap** [comment]
> @teicap what would you do if you were starting from a clean slate? Is the existing association setup good enough and it is "only" the management of existing connections when endpoints disappear that need to be fixed or is this just one detail in a larger problem?

@qabpea I see that the current support of SCTP is only a subset of what the protocol does and is not enough yet even for basic SCTP users. SCTP without multihoming is actually not SCTP.
On SCTP it's not the local node that decides on multihoming support, it's rather the peer that, either with the legacy support of IP addresses as embedded in the INIT/INIT-ACK chunks or with the ASCONF procedure.
Let's assume that you have a distributed SCTP Endpoint, that is you have 2 or more PODs that are sharing the same service in K8s terminology. One of these PODs will have a single-homed Association towards a remote peer. At a certain point the remote peer will ask to add another IP address with ASCONF procedure, this is embedded in the Association and may be encrypted by means of DTLS (rfc6083). There's a chance that the first message (probably an HB) will come from the remote peer to the tracker, that will not know what to do with that unknown message (unknown S-IP, same common D-IP). The tracker will decide to send that message towards one of the PODs with some criteria, if it fails choosing the host the HB will be sent to a POD that doesn't know about and will reply with ABORT (as this is seen as OOTB packet), such ABORT will force the remote peer to close the whole Association.

This means that, as it is today, it's not possible to instantiate replicas of SCTP Service unless you are 100% sure that the remote peer is single homed.

In this subset of SCTP we can talk about closing the service, that in my opinion should be a proper handling of SIGTERM, SIGKILL and SIGPIPE of the SCTP implementation and not part of the Transport, i.e. K8s CNI.

**Comment by thockin** [comment]
If I can offer a summary:

1) Lots of people do SCTP in userspace, which means that depending on a kernel-generated finalization sequence is dubious
2) Lots of people bind source port with SCTP
3) Leaving conntrack records which match the {protocol, client IP:port, service IP:port} will result in tears more often than not

As such, we should treat SCTP like UDP from a conntrack POV and clear tracking records.  This may result in timeouts for cases that use kernel-mode SCTP, but that seems like the lesser evil.

So what is left with this issue?  We probably need to audit code to make sure follow https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1071832988 or something like it as closely as possible.

I propose to rename it "Document and implement proper endpoint-lifecycle" or something like that.  

Anyone disagrees with the summary?

**Comment by qabpea** [comment]
I agree with your conclusions.

--Per
________________________________

**Comment by danwinship** [comment]
OK, so it's been a while and I don't remember a lot of the discussion here, but... it turns out we were arguing based on a fundmentally flawed premise... kube-proxy doesn't actually delete conntrack entries for SCTP services!

Yes, #74073 changed `pkg/proxy/iptables/proxier.go` to check "udp or sctp" rather than just "udp" when going through the list of "stale services" and "stale endpoints" whose conntrack entries need to be deleted... but in between when @uablrek filed the PR on Feb 14 2019 and when it merged on April 2 2020, @robscott snuck in with #83208 (merged on Oct 3, 2019) which changed the EndpointsMap code to limit the stale endpoints/services lists to UDP only (because that was all the proxiers cared about at that time). So now the iptables proxier code is checking "udp or sctp" on a list of services that has already been limited to udp-only. :man_facepalming:

(Lars mentioned in https://github.com/kubernetes/kubernetes/issues/101968#issuecomment-841791462 that "I actually did not test the previous fix myself, but the SCTP application developers discovering the problem did." and "it is not so easy for me to test". So presumably the original reporters tested his original fix but didn't retest it before it merged so the regression went unnoticed? And while in a perfect world this would have been caught by unit tests, the PR actually did update the appropriate unit test in `pkg/proxy/iptables/proxier_test.go` to test the SCTP case too... but unfortunately that test feeds objects of its own creation into `fp.deleteEndpointConnections()` rather than creating fake services and endpoints and pushing them through the fake proxy to get it to call that function itself. So it is just showing that `fp.deleteEndpointConnections()` will delete SCTP conntrack entries if you pass it data which `fp.syncProxyRules()` would never actually pass to it. :man_facepalming: :man_facepalming: )

(Noticed while writing [a patch to improve the naming of the conntrack-related fields in the update result objects](https://github.com/danwinship/kubernetes/commit/a5ab87dcd8d0c15a43aefd72eeff2c39ac468cc3), but now I don't want to push that PR since it seems like the current UDP-only status of those fields may be wrong? Or does the fact that we were actually _not_ flushing SCTP conntrack connections all this time mean that we actually don't need to be? Did this whole discussion get started because ovn-kubernetes copied what we _thought_ kube-proxy was doing, and then ran into problems, while kube-proxy itself never had any problem because it wasn't actually doing what we thought it was doing?  :man_facepalming: :man_facepalming: :man_facepalming:)

**Comment by tssurya** [comment]
> OK, so it's been a while and I don't remember a lot of the discussion here, but... it turns out we were arguing based on a fundmentally flawed premise... kube-proxy doesn't actually delete conntrack entries for SCTP services!
> 
> Yes, #74073 changed `pkg/proxy/iptables/proxier.go` to check "udp or sctp" rather than just "udp" when going through the list of "stale services" and "stale endpoints" whose conntrack entries need to be deleted... but in between when @uablrek filed the PR on Feb 14 2019 and when it merged on April 2 2020, @robscott snuck in with #83208 (merged on Oct 3, 2019) which changed the EndpointsMap code to limit the stale endpoints/services lists to UDP only (because that was all the proxiers cared about at that time). So now the iptables proxier code is checking "udp or sctp" on a list of services that has already been limited to udp-only. man_facepalming
> 
> (Lars mentioned in [#101968 (comment)](https://github.com/kubernetes/kubernetes/issues/101968#issuecomment-841791462) that "I actually did not test the previous fix myself, but the SCTP application developers discovering the problem did." and "it is not so easy for me to test". So presumably the original reporters tested his original fix but didn't retest it before it merged so the regression went unnoticed? And while in a perfect world this would have been caught by unit tests, the PR actually did update the appropriate unit test in `pkg/proxy/iptables/proxier_test.go` to test the SCTP case too... but unfortunately that test feeds objects of its own creation into `fp.deleteEndpointConnections()` rather than creating fake services and endpoints and pushing them through the fake proxy to get it to call that function itself. So it is just showing that `fp.deleteEndpointConnections()` will delete SCTP conntrack entries if you pass it data which `fp.syncProxyRules()` would never actually pass to it. man_facepalming man_facepalming )
> 

woah nice find on the regression!

> (Noticed while writing [a patch to improve the naming of the conntrack-related fields in the update result objects](https://github.com/danwinship/kubernetes/commit/a5ab87dcd8d0c15a43aefd72eeff2c39ac468cc3), but now I don't want to push that PR since it seems like the current UDP-only status of those fields may be wrong? Or does the fact that we were actually _not_ flushing SCTP conntrack connections all this time mean that we actually don't need to be? Did this whole discussion get started because ovn-kubernetes copied what we _thought_ kube-proxy was doing, and then ran into problems, while kube-proxy itself never had any problem because it wasn't actually doing what we thought it was doing? man_facepalming man_facepalming man_facepalming)

yes that's it! thankfully, ovn-k stuck to only deleting conntrack for UDP (https://github.com/ovn-org/ovn-kubernetes/pull/2829) and we ignore SCTP/TCP for endpoints deletion and endpoint slices deletion so unintentionally I guess we are now doing the same thing as kube-proxy.

**Comment by danwinship** [comment]
> but unfortunately that test feeds objects of its own creation into `fp.deleteEndpointConnections()` rather than creating fake services and endpoints and pushing them through the fake proxy to get it to call that function itself

I hacked up the test to push things through `fp.syncProxyRules()` (https://github.com/danwinship/kubernetes/commit/01b3ac121f21c50fc71fdbba82ec51200ca17bea) and confirmed that it no longer passes:

```
--- FAIL: TestDeleteEndpointConnectionsIPv4 (0.00s)
    proxier_test.go:212: V4 SCTP: Expected conntrack to be executed 1 times, but got 0
```

So... how is it that multiple people reported problems with SCTP services before #74073 but no one really has since then, despite the fact that it didn't actually do what it was supposed to do?

I guess looking more carefully, #74073 tried to do _two_ things:
1. Clear conntrack entries for SCTP in addition to UDP. (This part did not actually work.)
2. Change IPVS proxy to not do graceful termination of SCTP connections. (This part did work.)

It is possible that the second part was the important part all along? And is that specific to the differences in how iptables and ipvs work, so iptables didn't need any corresponding fix? Or is everyone using SCTP is also using IPVS?

**Comment by uablrek** [comment]
Without actually knowing, I think Telco abandoned K8s SCTP services (@teiclap?) since multihoming is not supported (and likely never will). Other users (or Telco in special cases) maybe don't bind the source port, and thus will not suffer from these problems. That would explain why no one has complained since 2020 even if the problem still exist.

**Comment by teiclap** [comment]
I wouldn't say Telco abandoned K8s SCTP services, it's rather the fact that K8s SCTP services as they are offered are not mature yet and need some further development.

**Comment by palonsoro** [comment]
> So... how is it that multiple people reported problems with SCTP services before https://github.com/kubernetes/kubernetes/pull/74073 but no one really has since then, despite the fact that it didn't actually do what it was supposed to do?

From what I have seen in the field, SCTP usage is not very common, so this may explain why it is hard to have bugs and/or regressions reported for it. 

> I wouldn't say Telco abandoned K8s SCTP services, it's rather the fact that K8s SCTP services as they are offered are not mature yet and need some further development.

That's correct from what I can see around. There are few people using SCTP, but not zero.

**Comment by ravens** [comment]
also I suspect many telco - as my current org -  are using multus to control the interface where SCTP is used - we for example usually need static IP and need a proper control of the fabric underneath so the interface is typically "mounted" in the pods via multus, bypassing many of the k8s SCTP services. 

At personal capacity, I still think the SCTP support need make sense at k8s level.

**Comment by qabpea** [comment]
I agree with Lars on this, SCTP usage in combination with the kube proxy must be quite limited, the service offered are not good enough for any actual use case I can think of.

//Per

**Comment by thockin** [comment]
> At personal capacity, I still think the SCTP support need make sense at k8s level.

I don't think anyone here disagrees, but sig-net lacks _active_ representation of this user-base.  Without a champion or champions, SCTP support will forever be a niche thing that isn't well supported.  I don't just need bug reports, but easy repros, test cases, and ideally hands-on-keyboards from people who know and live it.

I know TCP and UDP pretty well, and that covers 98% of today's k8s customers.  If we want to expand to new use-cases, we need some experts to chime in.  (and my thanks to those people who have!)

**Comment by teiclap** [comment]
One of the missing support for SCTP, but not limited to SCTP, is the possibility in K8s to choose what Source IP address and network to be used for egress traffic based on Service.
Calico finally allows that, but it must be a basic feature in k8s so that at least a basic traffic engineering is possible.
What is needed in such case from k8s-master is to create/update a static routing rule and a NAT rule saying that
IF any of the PODS implementing service S tried to send a PKT towards the external network
THEN change the Source IP with that : xx.xx.xx.xx 
 and set the static routing rule "If PKT has Source IP = xx.xx.xx.xx the default GW is YYY"

**Comment by aojea** [comment]
🙃

**Comment by aojea** [comment]
Since we now have more states on the EndpointSlices Ready, Serving and Terminating, I think that we can describe the lifecycle correctly. Something like:

**New connections** will be sent ONLY to **Ready endpoints**, IF  there are **NO Ready Endpoints** it will be sent to *Serving AND Terminating endpoints**

I think that this is clear, now the point of contention 😄 

**Existing connections** will be respected for **Ready OR Serving endpoints** 

We don't break connections to Terminating endpoints to allow graceful termination. And that leaves us with the state of a pod that is not ready or serving or is no longer part of the Service, if we want to implement active passive deployments we DO need to break these connections 

**Existing connections** will be closed for endpoints that are no longer part of the Service OR are **NOT Ready AND Terminating**

**Comment by thockin** [comment]
Re-stating from https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1071832988 and updating with your comments

For a given Service, with a given set of endpoints:

1a) New connections will be sent to ready (serving-and-not-terminating) endpoints, if any exist.
1b) If there are no ready endpoints, new connections will be sent to serving-and-terminating endpoints, if any exist.
1c) If there are no serving endpoints, new connections will be rejected.

2) When an endpoint changes from ready to not-ready (either not-serving, terminating, or both), existing connections will be retained.

3) When an endpoint is removed from the set, existing connections will be broken.

4) When an Endpoints or EndpointSlice resource is deleted (either actually deleted or the deletion timestamp set), endpoints in that resource will be removed from the set.

5) When the Service is deleted (either actually deleted or the deletion timestamp set), new connections will be ignored and existing connections will be broken.


Open: We need to use strong, well-defined words.  Probably most "will" -> "MUST" ?

Open: my (3) and (5) are not true today in k-p and I don't know how to make them true.  Perhaps we need to make those "MAY" or "SHOULD"?

Open: My (2) disagress with your assertion that "Existing connections will be closed for endpoints that are ... NOT Ready AND Terminating".  I am not confident in this.  Is there a reason to actively kill connections to not-serving pods?  Some people use pod readiness as a sort of "flow control" to stop incoming connections when they are busy.  We may need yet another flag here to indicate intention (assuming we have a mechanism).

**Comment by danwinship** [comment]
> New connections will be sent to ready (serving-and-not-terminating) endpoints

(As noted in https://github.com/kubernetes/kubernetes/pull/119394#discussion_r1285186623, this is not quite correct; the endpoints of a Service with `publishNotReadyAddresses` set are always `Ready` even when they aren't `Serving`. We were debating whether it would be more correct for them to be always `Serving` instead...)

> 2) When an endpoint changes from ready to not-ready (either not-serving, terminating, or both), existing connections will be retained.

We should probably explicitly define something like "the active endpoints", which is either the serving-non-terminating endpoints if there are any, or else the serving-terminating endpoints if not. Especially, because when the last serving-non-terminating endpoint terminates, it _doesn't_ stop being one of "the active endpoints", because at that point the rule for choosing "the active endpoints" flips...

> 3) When an endpoint is removed from the set...

Meaning what exactly? Only the case in (4)? When a Pod is deleted? When a Pod stops matching the `selector` of a Service?

> 4) When an Endpoints or EndpointSlice resource is deleted (either actually deleted or the deletion timestamp set), endpoints in that resource will be removed from the set.

kube-proxy does not look at Endpoints at all

> Open: my (3) and (5) are not true today in k-p and I don't know how to make them true. Perhaps we need to make those "MAY" or "SHOULD"?

We could use something very similar to the existing conntrack-UDP-fixing-up code. When a TCP endpoint IP leaves the set of all endpoints, call `conntrack -D` on any conntrack records mapping from the service IP to the deleted endpoint IP. When a Service is deleted, call `conntrack -D` on all records mapping from the service IP to anywhere.

**Comment by aojea** [comment]
> Open: My (2) disagress with your assertion that "Existing connections will be closed for endpoints that are ... NOT Ready AND Terminating"

I disagree with myself and agree with you, let the Pod process handle the connection, if the endpoint is present in the set he should own the connection

**Comment by thockin** [comment]
>> 3. When an endpoint is removed from the set...
>
> Meaning what exactly? Only the case in (4)? When a Pod is deleted? When a Pod stops matching the selector of a Service?

I meant this as "when a record is removed from the set of endpoints in the Endpoints resource or the set of EndpointSlice resources".  It's not making a statement about the actual workload, just the representation in Endpoints/EPSlice

> kube-proxy does not look at Endpoints at all

Yeah, but some simpler controllers might still.  We can spec this arbitrarily finely.  Define the "set of known endpoints" as either the set of records in the single Endpoints object OR the set of all records in all matching EndpointSlice objects, as determined by the controller (note that EndpointSlices should always be a superset of Endpoints)"

> When a TCP endpoint IP leaves the set of all endpoints, call conntrack -D on any conntrack records mapping from the service IP to the deleted endpoint IP

Try it.  Unless I screwed up all my testing, the kernel doesn't break the connection and in fact somehow recreates the conntrack.

**Comment by aojea** [comment]
I think we can implement the desired behavior of RST connections to Pods that are no longer part of the Service, we need to process the ESTABLISHED connection for the filter in the Forward hook, and also reject with a RST packet, test in https://lore.kernel.org/netfilter-devel/20250313231341.3040002-1-aojea@google.com/T/#u

If we can make clearly this distinction between, "Pods is being terminated" and "Pod no longer part of the Service", I think this will give us in a better place and also solve the long standing request of implementing Active / Passive deployments using Services and removing the Pods selector https://github.com/kubernetes/kubernetes/issues/45300 and move towards a more estandarized behavior across the ecosystem https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1694521909

This definitevely needs a KEP

Daman has stepped up to work on this, so I'm going to assign it to him
/assign @aroradaman

**Comment by aojea** [comment]
> Something like:
> 
> * put all valid endpoint IPs in an ipset
> * reject any established connections to IPs not in the set

heh, I see you noticed it, I realized my first test was with the endpoint IP and that can have all the issues you mention of not scaling well, if you see my second version of the patch we can filter by the Service ClusterIP https://lore.kernel.org/netfilter-devel/20250313231341.3040002-1-aojea@google.com/T/#m44c410cd56bdc62cc717798a0e28ee4dcaaab9ad , that simplifies everything and is something we are already doing, we have already reject rules for ClusterIPs that does not have endpoints

**Comment by thockin** [comment]
I didn't look at the patch, I just took an educated guess at where the most obvious traps might be. :)

**Comment by rikeshpatel14** [comment]
**Team** - Is anyone already actively working on this. If not, am interested in drafting a KEP and it's implementation. Can you please share contact details of individuals who can review the KEP. Is it below:

- Mailing List: kubernetes-sig-network@googlegroups.com
- Slack: #sig-network
